File: 131-palindrome-partitioning/palindrome-partitioning.java
class Solution {
    public List<List<String>> partition(String s) {
        List<List<String>> res = new ArrayList<>();
        List<String> ans = new ArrayList<>();
        String current = "";
        allWays(s, res, ans, current, 0);
        return res;
    }

    public static void allWays(String s, List<List<String>> res, List<String> ans, String current, int idx) {
        if (idx == s.length()) {
            if (current.length() == 0) {
                res.add(new ArrayList<>(ans));
            }
            return;
        }

        current += s.charAt(idx);
        if (isValidPalindrome(current)) {
            ans.add(current);
            allWays(s, res, ans, "", idx + 1);
            ans.remove(ans.size() - 1);
        }

        allWays(s, res, ans, current, idx + 1);
    }

    public static boolean isValidPalindrome(String str) {
        int l = 0, r = str.length() - 1;
        while (l < r) {
            if (str.charAt(l++) != str.charAt(r--)) return false;
        }
        return true;
    }
}

File: 131-palindrome-partitioning/palindrome-partitioning.cpp
class Solution {
public:
    vector<vector<string>>res;
    bool isPal(string s, int st, int end){
        while(st<=end){
            if(s[st++] != s[end--])
                return false;
        }
        return true;
    }
    void helper(string s, int idx, vector<string>ans){

        if(idx == s.length())
        {
            res.push_back(ans);
            return;
        }

        for(int i=idx; i<s.length(); i++){
            if(isPal(s, idx, i)){
                ans.push_back(s.substr(idx, i-idx+1));
                helper(s, i+1, ans);
                ans.pop_back();
            }
        }
    }
    vector<vector<string>> partition(string s) {
       vector<string>ans;
       helper(s, 0, ans);
       return res;
     
    }
};